<?xml version="1.0" ?>
<robot name="drone_frame" xmlns:xacro="http://www.ros.org/wiki/xacro">

<xacro:include filename="$(find drone_frame_description)/urdf/materials.xacro" />

<!-- Base link -->
<link name="base_link">
  <inertial>
    <origin xyz="5.361940941515797e-15 -0.006931498212032427 1.0942736615338362e-16" rpy="0 0 0"/>
    <mass value="0.950824254848482"/>
    <inertia ixx="0.010991" iyy="0.017094" izz="0.010991" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/base_link.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/base_link.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<!-- Arm links -->
<link name="arm_8_1">
  <inertial>
    <origin xyz="-0.0973299825380552 2.527436940291672e-11 -0.08302889455495804" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.0006" iyy="0.000929" izz="0.000389" ixy="0.0" iyz="0.0" ixz="-0.000277"/>
  </inertial>
  <visual>
    <origin xyz="0.2 -0.02 0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_8_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="0.2 -0.02 0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_8_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_7_1">
  <inertial>
    <origin xyz="-0.08302889455495802 -2.527437287236367e-11 -0.0973299825380552" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.000389" iyy="0.000929" izz="0.0006" ixy="-0.0" iyz="-0.0" ixz="-0.000277"/>
  </inertial>
  <visual>
    <origin xyz="0.055 -0.02 0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_7_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="0.055 -0.02 0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_7_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_6_1">
  <inertial>
    <origin xyz="0.09732998253805486 2.5274376341810623e-11 0.08302889455495802" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.0006" iyy="0.000929" izz="0.000389" ixy="-0.0" iyz="-0.0" ixz="-0.000277"/>
  </inertial>
  <visual>
    <origin xyz="-0.2 -0.02 -0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_6_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="-0.2 -0.02 -0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_6_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_5_1">
  <inertial>
    <origin xyz="0.08302889455495802 -2.527437287236367e-11 0.0973299825380552" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.000389" iyy="0.000929" izz="0.0006" ixy="0.0" iyz="0.0" ixz="-0.000277"/>
  </inertial>
  <visual>
    <origin xyz="-0.055 -0.02 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_5_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="-0.055 -0.02 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_5_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_4_1">
  <inertial>
    <origin xyz="0.0973299825380552 -2.527437287236367e-11 -0.08302889455495804" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.0006" iyy="0.000929" izz="0.000389" ixy="0.0" iyz="-0.0" ixz="0.000277"/>
  </inertial>
  <visual>
    <origin xyz="-0.2 -0.02 0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_4_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="-0.2 -0.02 0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_4_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_3_1">
  <inertial>
    <origin xyz="0.08302889455495802 2.527436940291672e-11 -0.0973299825380552" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.000389" iyy="0.000929" izz="0.0006" ixy="-0.0" iyz="0.0" ixz="0.000277"/>
  </inertial>
  <visual>
    <origin xyz="-0.055 -0.02 0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_3_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="-0.055 -0.02 0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_3_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_2_1">
  <inertial>
    <origin xyz="-0.09732998253804703 -2.5274383280704527e-11 0.08302889455496801" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.0006" iyy="0.000929" izz="0.000389" ixy="-0.0" iyz="0.0" ixz="0.000277"/>
  </inertial>
  <visual>
    <origin xyz="0.2 -0.02 -0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="0.2 -0.02 -0.055" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="arm_1_1">
  <inertial>
    <origin xyz="-0.08302889455495796 2.5278622944879814e-11 0.0973299825380552" rpy="0 0 0"/>
    <mass value="0.15422742956922936"/>
    <inertia ixx="0.000389" iyy="0.000929" izz="0.0006" ixy="0.0" iyz="-0.0" ixz="0.000277"/>
  </inertial>
  <visual>
    <origin xyz="0.055 -0.02 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <origin xyz="0.055 -0.02 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/arm_1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<!-- Holder links -->
<link name="holder_4">
  <inertial>
    <origin xyz="0.0012743502456399602 -1.933707013956898e-11 -0.016274350245639946" rpy="0 0 0"/>
    <mass value="0.545696208866394"/>
    <inertia ixx="0.00016" iyy="0.000161" izz="0.00016" ixy="-0.0" iyz="0.0" ixz="-1e-06"/>
  </inertial>
  <visual>
    <origin xyz="0.215 -0.02 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_4_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="0.2 -0.02 -0.215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_4_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="holder_3">
  <inertial>
    <origin xyz="0.01627435024563989 3.173676564605099e-10 0.0012743502456399047" rpy="0 0 0"/>
    <mass value="0.545696208866394"/>
    <inertia ixx="0.00016" iyy="0.000161" izz="0.00016" ixy="0.0" iyz="0.0" ixz="1e-06"/>
  </inertial>
  <visual>
    <origin xyz="0.20 -0.02 0.215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_3_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="0.215 -0.02 0.215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_3_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="holder_2">
  <inertial>
    <origin xyz="-0.0012743502456399047 -1.9344432306001025e-11 0.01627435024563989" rpy="0 0 0"/>
    <mass value="0.545696208866394"/>
    <inertia ixx="0.00016" iyy="0.000161" izz="0.00016" ixy="0.0" iyz="-0.0" ixz="-1e-06"/>
  </inertial>
  <visual>
    <origin xyz="-0.215 -0.02 0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="-0.2 -0.02 0.215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="holder_1">
  <inertial>
    <origin xyz="-0.01627435024563989 -1.9344414958766265e-11 -0.0012743502456399047" rpy="0 0 0"/>
    <mass value="0.545696208866394"/>
    <inertia ixx="0.00016" iyy="0.000161" izz="0.00016" ixy="0.0" iyz="0.0" ixz="1e-06"/>
  </inertial>
  <visual>
    <origin xyz="-0.2 -0.02 -0.215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="-0.215 -0.02 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="file://$(find drone_frame_description)/meshes/holder_1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<!-- Virtual intermediate links for redundant actuation -->
<link name="virtual_1">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <mass value="0.001"/>
    <inertia ixx="0.000001" iyy="0.000001" izz="0.000001" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
    <material name="transparent"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="virtual_2">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <mass value="0.001"/>
    <inertia ixx="0.000001" iyy="0.000001" izz="0.000001" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
    <material name="transparent"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="virtual_3">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <mass value="0.001"/>
    <inertia ixx="0.000001" iyy="0.000001" izz="0.000001" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
    <material name="transparent"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="virtual_4">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <mass value="0.001"/>
    <inertia ixx="0.000001" iyy="0.000001" izz="0.000001" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
    <material name="transparent"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<!-- Main actuation joints from base link to arms -->
<joint name="revolute_1" type="revolute">
  <origin xyz="-0.2 0.02 -0.055" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_8_1"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_2" type="revolute">
  <origin xyz="-0.055 0.02 -0.2" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_7_1"/>
  <axis xyz="1.0 0.0 0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_3" type="revolute">
  <origin xyz="0.2 0.02 0.055" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_6_1"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_4" type="revolute">
  <origin xyz="0.055 0.02 0.2" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_5_1"/>
  <axis xyz="1.0 0.0 0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_5" type="revolute">
  <origin xyz="0.2 0.02 -0.055" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_4_1"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_6" type="revolute">
  <origin xyz="0.055 0.02 -0.2" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_3_1"/>
  <axis xyz="1.0 0.0 0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_7" type="revolute">
  <origin xyz="-0.2 0.02 0.055" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_2_1"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_8" type="revolute">
  <origin xyz="-0.055 0.02 0.2" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="arm_1_1"/>
  <axis xyz="1.0 0.0 0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<!-- Virtual linking joints -->
<joint name="revolute_9" type="revolute">
  <origin xyz="-0.16 0.0 0.0" rpy="0 0 0"/>
  <parent link="arm_1_1"/>
  <child link="virtual_4"/>
  <axis xyz="0.0 0.0 -1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_10" type="revolute">
  <origin xyz="-0.015 0.0 0.145" rpy="0 0 0"/>
  <parent link="arm_2_1"/>
  <child link="virtual_4"/>
  <axis xyz="1.0 0.0 -0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_11" type="revolute">
  <origin xyz="0.0 0.0 -0.16" rpy="0 0 0"/>
  <parent link="arm_8_1"/>
  <child link="virtual_3"/>
  <axis xyz="1.0 0.0 0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_12" type="revolute">
  <origin xyz="-0.145 0.0 -0.015" rpy="0 0 0"/>
  <parent link="arm_7_1"/>
  <child link="virtual_3"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_13" type="revolute">
  <origin xyz="0.16 0.0 0.0" rpy="0 0 0"/>
  <parent link="arm_3_1"/>
  <child link="virtual_2"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_14" type="revolute">
  <origin xyz="0.015 0.0 -0.145" rpy="0 0 0"/>
  <parent link="arm_4_1"/>
  <child link="virtual_2"/>
  <axis xyz="-1.0 0.0 0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_15" type="revolute">
  <origin xyz="0.0 0.0 0.16" rpy="0 0 0"/>
  <parent link="arm_6_1"/>
  <child link="virtual_1"/>
  <axis xyz="-1.0 -0.0 -0.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<joint name="revolute_16" type="revolute">
  <origin xyz="0.145 0.0 0.015" rpy="0 0 0"/>
  <parent link="arm_5_1"/>
  <child link="virtual_1"/>
  <axis xyz="0.0 0.0 -1.0"/>
  <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
</joint>

<!-- Fixed joints connecting virtual links to holders -->
<joint name="rigid_17" type="fixed">
  <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  <parent link="virtual_1"/>
  <child link="holder_1"/>
</joint>

<joint name="rigid_18" type="fixed">
  <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  <parent link="virtual_2"/>
  <child link="holder_2"/>
</joint>

<joint name="rigid_19" type="fixed">
  <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  <parent link="virtual_3"/>
  <child link="holder_3"/>
</joint>

<joint name="rigid_20" type="fixed">
  <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  <parent link="virtual_4"/>
  <child link="holder_4"/>
</joint>

</robot>
